pvanalytics/quality/irradiance.py

```diff
@@ -1 +1,333 @@
 """Quality control functions for irradiance data."""
+
+import numpy as np
+from pvlib.tools import cosd
+
+
+QCRAD_LIMITS = {'ghi_ub': {'mult': 1.5, 'exp': 1.2, 'min': 100},
+                'dhi_ub': {'mult': 0.95, 'exp': 1.2, 'min': 50},
+                'dni_ub': {'mult': 1.0, 'exp': 0.0, 'min': 0},
+                'ghi_lb': -4, 'dhi_lb': -4, 'dni_lb': -4}
+
+QCRAD_CONSISTENCY = {
+    'ghi_ratio': {
+        'low_zenith': {
+            'zenith_bounds': [0.0, 75],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.92, 1.08]},
+        'high_zenith': {
+            'zenith_bounds': [75, 93],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.85, 1.15]}},
+    'dhi_ratio': {
+        'low_zenith': {
+            'zenith_bounds': [0.0, 75],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.0, 1.05]},
+        'high_zenith': {
+            'zenith_bounds': [75, 93],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.0, 1.10]}}}
+
+
+def _check_limits(val, lb=None, ub=None, lb_ge=False, ub_le=False):
+    """Return True where lb < (or <=) val < (or <=) ub."""
+    if lb_ge:
+        lb_op = np.greater_equal
+    else:
+        lb_op = np.greater
+    if ub_le:
+        ub_op = np.less_equal
+    else:
+        ub_op = np.less
+
+    if (lb is not None) & (ub is not None):
+        return lb_op(val, lb) & ub_op(val, ub)
+    elif lb is not None:
+        return lb_op(val, lb)
+    elif ub is not None:
+        return ub_op(val, ub)
+    else:
+        raise ValueError('must provide either upper or lower bound')
+
+
+def _qcrad_ub(dni_extra, sza, lim):
+    cosd_sza = cosd(sza)
+    cosd_sza[cosd_sza < 0] = 0
+    return lim['mult'] * dni_extra * cosd_sza**lim['exp'] + lim['min']
+
+
+def check_ghi_limits_qcrad(ghi, solar_zenith, dni_extra, limits=None):
+    r"""Test for physical limits on GHI using the QCRad criteria.
+
+    Test passes if a value > lower bound and value < upper bound. Lower bounds
+    are constant for all tests. Upper bounds are calculated as
+
+    .. math::
+        ub = min + mult * dni\_extra * cos( solar\_zenith)^{exp}
+
+    Parameters
+    ----------
+    ghi : Series
+        Global horizontal irradiance in :math:`W/m^2`
+    solar_zenith : Series
+        Solar zenith angle in degrees
+    dni_extra : Series
+        Extraterrestrial normal irradiance in :math:`W/m^2`
+    limits : dict, default QCRAD_LIMITS
```

Comment:
This function tests only GHI, should we instead describe a dict with the `ghi_ub` and `ghi_lb` keys? Could still be QCRAD_LIMITS by default.

Reply:
This seems reasonable. We should do the same for `check_[dni/dhi]_limits`